File: ios-oc/VoiceAgent/Chat/TranscriptMessageCell.m
//
//  TranscriptMessageCell.m
//  VoiceAgent
//
//  Created by qinhui on 2025/11/17.
//

#import "TranscriptMessageCell.h"
#import <Masonry/Masonry.h>
#import "VoiceAgent-Swift.h"

static inline UIColor *VAHexColor(NSUInteger hexValue, CGFloat alpha) {
    return [UIColor colorWithRed:((hexValue >> 16) & 0xFF) / 255.0
                           green:((hexValue >> 8) & 0xFF) / 255.0
                            blue:(hexValue & 0xFF) / 255.0
                           alpha:alpha];
}

@interface TranscriptMessageCell ()

@property (nonatomic, strong) UIView *avatarView;
@property (nonatomic, strong) UILabel *avatarLabel;
@property (nonatomic, strong) UILabel *messageLabel;
@property (nonatomic, strong) UIView *bubbleView;
@property (nonatomic, strong) MASConstraint *bubbleLeading;
@property (nonatomic, strong) MASConstraint *bubbleTrailing;
@property (nonatomic, strong) MASConstraint *avatarLeading;
@property (nonatomic, strong) MASConstraint *avatarTrailing;

@end

@implementation TranscriptMessageCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setupUI {
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.backgroundColor = [UIColor clearColor];
    self.contentView.backgroundColor = [UIColor clearColor];
    
    self.avatarView = [[UIView alloc] init];
    self.avatarView.layer.cornerRadius = 16;
    [self.contentView addSubview:self.avatarView];
    
    self.avatarLabel = [[UILabel alloc] init];
    self.avatarLabel.font = [UIFont systemFontOfSize:12 weight:UIFontWeightSemibold];
    self.avatarLabel.textColor = [UIColor whiteColor];
    self.avatarLabel.textAlignment = NSTextAlignmentCenter;
    [self.avatarView addSubview:self.avatarLabel];
    
    self.bubbleView = [[UIView alloc] init];
    self.bubbleView.layer.cornerRadius = 16;
    [self.contentView addSubview:self.bubbleView];

    self.messageLabel = [[UILabel alloc] init];
    self.messageLabel.font = [UIFont systemFontOfSize:14];
    self.messageLabel.numberOfLines = 0;
    [self.bubbleView addSubview:self.messageLabel];
    
    [self.avatarView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.contentView).offset(4);
        make.width.height.mas_equalTo(32);
        self.avatarLeading = make.left.equalTo(self.contentView).offset(16);
        self.avatarTrailing = make.right.equalTo(self.contentView).offset(-16);
    }];
    [self.avatarTrailing deactivate];
    
    [self.avatarLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(self.avatarView);
    }];
    
    [self.bubbleView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.contentView).offset(4);
        make.bottom.equalTo(self.contentView).offset(-4);
        make.width.lessThanOrEqualTo(self.contentView).multipliedBy(0.75);
        self.bubbleLeading = make.left.equalTo(self.avatarView.mas_right).offset(8);
        self.bubbleTrailing = make.right.equalTo(self.avatarView.mas_left).offset(-8);
    }];
    [self.bubbleTrailing deactivate];

    [self.messageLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.bubbleView).inset(12);
    }];
}

- (void)configureWithTranscript:(id)transcript {
    Transcript *t = (Transcript *)transcript;
    BOOL isAgent = t.type == TranscriptTypeAgent;
    self.avatarView.backgroundColor = isAgent ? VAHexColor(0x3B82F6, 1.0) : VAHexColor(0x10B981, 1.0);
    self.avatarLabel.text = isAgent ? @"AI" : @"Me";
    self.bubbleView.backgroundColor = isAgent ? VAHexColor(0x334155, 1.0) : VAHexColor(0x2563EB, 1.0);
    self.messageLabel.textColor = isAgent ? VAHexColor(0xF1F5F9, 1.0) : [UIColor whiteColor];
    self.messageLabel.text = t.text;

    if (isAgent) {
        [self.avatarTrailing deactivate];
        [self.bubbleTrailing deactivate];
        [self.avatarLeading activate];
        [self.bubbleLeading activate];
    } else {
        [self.avatarLeading deactivate];
        [self.bubbleLeading deactivate];
        [self.avatarTrailing activate];
        [self.bubbleTrailing activate];
    }
}

@end

File: ios-swift/VoiceAgent/Chat/TranscriptMessageCell.swift
//
//  TranscriptMessageCell.swift
//  VoiceAgent
//

import UIKit
import SnapKit

class TranscriptMessageCell: UITableViewCell {
    static let reuseIdentifier = "TranscriptMessageCell"

    private let avatarView = UIView()
    private let avatarLabel = UILabel()
    private let bubbleView = UIView()
    private let messageLabel = UILabel()

    private var bubbleLeading: Constraint?
    private var bubbleTrailing: Constraint?
    private var avatarLeading: Constraint?
    private var avatarTrailing: Constraint?

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupUI()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func setupUI() {
        selectionStyle = .none
        backgroundColor = .clear
        contentView.backgroundColor = .clear

        avatarView.layer.cornerRadius = 16
        contentView.addSubview(avatarView)

        avatarLabel.font = .systemFont(ofSize: 11, weight: .semibold)
        avatarLabel.textColor = .white
        avatarLabel.textAlignment = .center
        avatarView.addSubview(avatarLabel)

        bubbleView.layer.cornerRadius = 16
        contentView.addSubview(bubbleView)

        messageLabel.font = .systemFont(ofSize: 14)
        messageLabel.numberOfLines = 0
        bubbleView.addSubview(messageLabel)

        avatarView.snp.makeConstraints { make in
            make.top.equalToSuperview().offset(4)
            make.width.height.equalTo(32)
            avatarLeading = make.left.equalToSuperview().offset(16).constraint
            avatarTrailing = make.right.equalToSuperview().offset(-16).constraint
        }
        avatarTrailing?.deactivate()

        avatarLabel.snp.makeConstraints { make in
            make.center.equalToSuperview()
        }

        bubbleView.snp.makeConstraints { make in
            make.top.equalToSuperview().offset(4)
            make.bottom.equalToSuperview().offset(-4)
            make.width.lessThanOrEqualToSuperview().multipliedBy(0.75)
            bubbleLeading = make.left.equalTo(avatarView.snp.right).offset(8).constraint
            bubbleTrailing = make.right.equalTo(avatarView.snp.left).offset(-8).constraint
        }
        bubbleTrailing?.deactivate()

        messageLabel.snp.makeConstraints { make in
            make.edges.equalToSuperview().inset(12)
        }
    }

    func configure(with transcript: Transcript) {
        let isAgent = transcript.type == .agent

        avatarView.backgroundColor = isAgent ? AppColors.avatarAgent : AppColors.avatarUser
        avatarLabel.text = isAgent ? "AI" : "Me"

        bubbleView.backgroundColor = isAgent ? AppColors.bubbleAgentBg : AppColors.bubbleUserBg
        messageLabel.textColor = isAgent ? AppColors.bubbleAgentText : AppColors.bubbleUserText
        messageLabel.text = transcript.text

        if isAgent {
            avatarLeading?.activate()
            avatarTrailing?.deactivate()
            bubbleLeading?.activate()
            bubbleTrailing?.deactivate()
            bubbleView.semanticContentAttribute = .forceLeftToRight
        } else {
            avatarLeading?.deactivate()
            avatarTrailing?.activate()
            bubbleLeading?.deactivate()
            bubbleTrailing?.activate()
            bubbleView.semanticContentAttribute = .forceRightToLeft
        }

        let smallCorner: CGFloat = 2
        let bigCorner: CGFloat = 16
        if isAgent {
            applyBubbleCorners(topLeft: smallCorner, topRight: bigCorner, bottomLeft: bigCorner, bottomRight: bigCorner)
        } else {
            applyBubbleCorners(topLeft: bigCorner, topRight: smallCorner, bottomLeft: bigCorner, bottomRight: bigCorner)
        }
    }

    private func applyBubbleCorners(topLeft: CGFloat, topRight: CGFloat, bottomLeft: CGFloat, bottomRight: CGFloat) {
        let path = UIBezierPath()
        bubbleView.layoutIfNeeded()
        let bounds = bubbleView.bounds.isEmpty ? CGRect(x: 0, y: 0, width: 200, height: 40) : bubbleView.bounds

        path.move(to: CGPoint(x: topLeft, y: 0))
        path.addLine(to: CGPoint(x: bounds.width - topRight, y: 0))
        path.addArc(withCenter: CGPoint(x: bounds.width - topRight, y: topRight), radius: topRight, startAngle: -.pi / 2, endAngle: 0, clockwise: true)
        path.addLine(to: CGPoint(x: bounds.width, y: bounds.height - bottomRight))
        path.addArc(withCenter: CGPoint(x: bounds.width - bottomRight, y: bounds.height - bottomRight), radius: bottomRight, startAngle: 0, endAngle: .pi / 2, clockwise: true)
        path.addLine(to: CGPoint(x: bottomLeft, y: bounds.height))
        path.addArc(withCenter: CGPoint(x: bottomLeft, y: bounds.height - bottomLeft), radius: bottomLeft, startAngle: .pi / 2, endAngle: .pi, clockwise: true)
        path.addLine(to: CGPoint(x: 0, y: topLeft))
        path.addArc(withCenter: CGPoint(x: topLeft, y: topLeft), radius: topLeft, startAngle: .pi, endAngle: -.pi / 2, clockwise: true)
        path.close()

        let mask = CAShapeLayer()
        mask.path = path.cgPath
        bubbleView.layer.mask = mask
    }

    override func prepareForReuse() {
        super.prepareForReuse()
        avatarLeading?.deactivate()
        avatarTrailing?.deactivate()
        bubbleLeading?.deactivate()
        bubbleTrailing?.deactivate()
        avatarLeading?.activate()
        bubbleLeading?.activate()
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        if let text = avatarLabel.text {
            let isAgent = text == "AI"
            let smallCorner: CGFloat = 2
            let bigCorner: CGFloat = 16
            if isAgent {
                applyBubbleCorners(topLeft: smallCorner, topRight: bigCorner, bottomLeft: bigCorner, bottomRight: bigCorner)
            } else {
                applyBubbleCorners(topLeft: bigCorner, topRight: smallCorner, bottomLeft: bigCorner, bottomRight: bigCorner)
            }
        }
    }
}
